src/Microsoft.Azure.SignalR/DependencyInjectionExtensions.cs

```diff
@@ -1,72 +1,73 @@
+        /// </summary>
+        /// <param name="builder">The <see cref="ISignalRServerBuilder"/>.</param>
+        /// <returns>The same instance of the <see cref="ISignalRServerBuilder"/> for chaining.</returns>
+        public static ISignalRServerBuilder AddAzureSignalR(this ISignalRServerBuilder builder)
+        {
+            builder.Services.AddSingleton<IConfigureOptions<ServiceOptions>, ServiceOptionsSetup>();
+            return builder.AddAzureSignalRCore();
+        }
+
+        /// <summary>
+        /// Adds the minimum essential Azure SignalR services to the specified <see cref="ISignalRServerBuilder" />.
+        /// </summary>
+        /// <param name="builder">The <see cref="ISignalRServerBuilder"/>.</param>
+        /// <param name="connectionString">The connection string of an Azure SignalR Service instance.</param>
+        /// <returns>The same instance of the <see cref="ISignalRServerBuilder"/> for chaining.</returns>
+        public static ISignalRServerBuilder AddAzureSignalR(this ISignalRServerBuilder builder, string connectionString)
+        {
+            return builder.AddAzureSignalR(options =>
+            {
+                options.ConnectionString = connectionString;
+            });
+        }
+
+        /// <summary>
+        /// Adds the minimum essential Azure SignalR services to the specified <see cref="ISignalRServerBuilder" />.
+        /// </summary>
+        /// <param name="builder">The <see cref="ISignalRServerBuilder"/>.</param>
+        /// <param name="configure">A callback to configure the <see cref="ServiceOptions"/>.</param>
+        /// <returns>The same instance of the <see cref="ISignalRServerBuilder"/> for chaining.</returns>
+        public static ISignalRServerBuilder AddAzureSignalR(this ISignalRServerBuilder builder, Action<ServiceOptions> configure)
+        {
+            builder.AddAzureSignalR()
+                   .Services.Configure(configure);
+
+            return builder;
+        }
+
+        private static ISignalRServerBuilder AddAzureSignalRCore(this ISignalRServerBuilder builder)
+        {
+            builder.Services
+                .AddSingleton(typeof(HubLifetimeManager<>), typeof(ServiceLifetimeManager<>))
+                .AddSingleton(typeof(IServiceProtocol), typeof(ServiceProtocol))
+                .AddSingleton(typeof(IClientConnectionManager), typeof(ClientConnectionManager))
+                .AddSingleton(typeof(IServiceConnectionManager<>), typeof(ServiceConnectionManager<>))
+                .AddSingleton(typeof(IServiceEndpointProvider), typeof(ServiceEndpointProvider))
+                .AddSingleton(typeof(ServiceHubDispatcher<>))
+                .AddSingleton(typeof(AzureSignalRMarkerService))
+                .AddSingleton<IClientConnectionFactory, ClientConnectionFactory>()
+                .AddSingleton<IHostedService, HeartBeat>()
+                .AddSingleton<NegotiateHandler>();

```

Comment:
This is the updated line. Not sure why the whole file is marked as updated.

Reply:
Are any of these services extension points for users? If so they should be `TryAddSingleton`

Reply:
Good point. In our case, none of them are extensible for user. All of them are internal, except two of them are public class/interface.
- `class HubLifetimeManager` is `public` and `abstract`. The implementation of `ServiceLifetimeManager` is service specific. So we don't want it to be overridden.
- `interface IServiceProtocol` is `public`. We don't want it to be overridden either.